File: pkg/conversation/spec/conversation_types.go
package spec

import (
	"time"
)

// ConversationRoleEnum represents the role of a participant in a conversation.
type ConversationRoleEnum string

const (
	System    ConversationRoleEnum = "system"
	User      ConversationRoleEnum = "user"
	Assistant ConversationRoleEnum = "assistant"
	Function  ConversationRoleEnum = "function"
	Feedback  ConversationRoleEnum = "feedback"
)

// ConversationMessage represents a message in a conversation.
type ConversationMessage struct {
	ID        string               `json:"id"`
	CreatedAt *time.Time           `json:"createdAt,omitempty"`
	Role      ConversationRoleEnum `json:"role"`
	Content   string               `json:"content"`
	Timestamp *string              `json:"timestamp,omitempty"`
	Name      *string              `json:"name,omitempty"`
	Details   *string              `json:"details,omitempty"`
}

// ConversationItem represents a conversation with basic details.
type ConversationItem struct {
	ID        string    `json:"id"`
	Title     string    `json:"title"`
	CreatedAt time.Time `json:"createdAt"`
}

// Conversation represents a conversation with messages.
type Conversation struct {
	ConversationItem
	ModifiedAt time.Time             `json:"modifiedAt"`
	Messages   []ConversationMessage `json:"messages"`
}

// IConversationAPI defines the interface for conversation-related operations.
type IConversationAPI interface {
	SaveConversation(conversation Conversation) error
	DeleteConversation(id string, title string) error
	GetConversation(id string, title string) (*Conversation, error)
	ListConversations(token *string) ([]ConversationItem, *string, error)
	AddMessageToConversation(id string, title string, newMessage ConversationMessage) error
}

File: frontend/app/models/conversationmodel/conversation_types.ts
export enum ConversationRoleEnum {
	system = 'system',
	user = 'user',
	assistant = 'assistant',
	function = 'function',
	feedback = 'feedback',
}
export interface ConversationMessage {
	id: string;
	createdAt?: Date;
	role: ConversationRoleEnum;
	content: string;
	timestamp?: string;
	name?: string;
	details?: string;
}

export interface ConversationItem {
	id: string;
	title: string;
	createdAt: Date;
}

export type Conversation = ConversationItem & {
	modifiedAt: Date;
	messages: ConversationMessage[];
};

export interface IConversationAPI {
	saveConversation: (conversation: Conversation) => Promise<void>;
	deleteConversation: (id: string, title: string) => Promise<void>;
	getConversation: (id: string, title: string) => Promise<Conversation | null>;
	listConversations: (token?: string) => Promise<{ conversations: ConversationItem[]; nextToken?: string }>;
	addMessageToConversation(id: string, title: string, newMessage: ConversationMessage): Promise<void>;
}
